File: lexical_analyzer.cpp
#include<iostream>
#include<fstream>
#include <iomanip>
#include <string>
#include <string.h>

using namespace std;

char operators[] = "*+-=/><%";
char separator[] = "'(){}[],:;";
char keywords[18][10] = { "int", "float", "bool", "if", "else", "then", "endif", "while", "whileend", "do", "doend","for", "forend", "input", "output", "and", "or", "function" };
bool hasSep = false;	// check if has separater
bool hasOpe = false;	// check if has operator
int currentIndex = 0;	// index for check isNumber() and isReal()

// FSM states for print
bool endOperator = false;
bool endSeparator = false;
bool endNumber = false;
bool endReal = false;
bool endIdentifier = false;

ifstream fin;		//read file
ofstream outputFile;
char ch;	// read file and get every character
char buffer[15];	//store characters in an array to determine isKeyword or isNumber or isReal or isIdentifier
int j = 0;		// index for buffer[]
bool isComments = false;	//check if it is comments
char detection[2] = { '\0' };	// detect '!' twice, if detect, add '!' to array according to occur times
int x = 0;	// used to counting detection array

bool isKeyword(char buffer[]) {
	for (int i = 0; i < 18; ++i)
		if (strcmp(keywords[i], buffer) == 0)
			return true;
	return false;
}

bool isOperator(char ch) {
	for (int i = 0; i < 10; ++i) {
		if (ch == operators[i])
			return true;
	}
	return false;	
}

bool isSeparator(char ch) {
	for (int i = 0; i < 10; ++i) {	//check separator
		if (ch == separator[i])
			return true;
	}
	return false;
}

bool isReal(char buffer[]) {
	string str(buffer);
	string substr(".");

	size_t found = str.find(substr);

	if (found != string::npos) {
		for (int i = 0; i < found; i++) {
			if (!isdigit(buffer[i]))
				return false;
			continue;
		}
		for (int i = found + 1; i < currentIndex; i++) {
			if (!isdigit(buffer[i]))
				return false;
			continue;
		}
		return true;
	}
	else
		return false;
}

bool isNumber(char buffer[]) {
	for (int i = 0; i <= (currentIndex - 1); i++) {
		if (!isdigit(buffer[i]))
			return false;
		continue;
	}
	return true;
}

void print() {

	if (endIdentifier) {
		cout << "IDENTIFIER      =     " << buffer << endl;
		outputFile << "IDENTIFIER      =     " << buffer << endl;
		endIdentifier = false;
	}

	if (endNumber) {
		cout << "NUMBER          =     " << buffer << endl;
		outputFile << "NUMBER          =     " << buffer << endl;

		endNumber = false;
	}

	if (endReal) {
		cout << "REAL            =     " << buffer << endl;
		outputFile << "REAL            =     " << buffer << endl;
		endReal = false;
	}

	if (endSeparator) {	// single separator detected, print it
		cout << "SEPARATOR       =     " << ch << endl;
		outputFile << "SEPARATOR       =     " << ch << endl;
		endSeparator = false;
		hasSep = false;
	}
	if (endOperator) { // single operator detected, print it
		cout << "OPERATOR        =     " << ch << endl;
		outputFile << "OPERATOR        =     " << ch << endl;
		endOperator = false;
		hasOpe = false;
	}
}

void lexer() {

	if (ch == '!') {	//check if is comment
		detection[x] = ch;
		x++;
		if (detection[1] == ch) {
			isComments = false;
			x = 0;
			detection[0] = '\0';
			detection[1] = '\0';
		}
		else
			isComments = true;
	}
	if (!isComments) {	// if not comment, start lexical analyzer


		if (isSeparator(ch)) {
			hasSep = true;
			endSeparator = true;
		}

		if (isOperator(ch)) {
			hasOpe = true;
			endOperator = true;
		}

		if (isalnum(ch) || ch == '$' || ch == '.') {	// check ch is a number or letter or '$' or '.'
			buffer[j++] = ch;		// if so, add to buffer
			currentIndex++;
		}
		else if ((ch == ' ' || ch == '\n' || hasSep || hasOpe || fin.eof()) && (j != 0)) {	// end of variables by detecting if j(index for buffer[]) != 0
			buffer[j] = '\0';	// set last char in buffer( might be ' ', '\n', separator, operator
			j = 0;	// reset the index for buffer[]
			if (isKeyword(buffer)) {	// check if the variable is keyword
				cout << "KEYWORD         =     " << buffer << endl;
				outputFile << "KEYWORD         =     " << buffer << endl;
			}
			else if (isReal(buffer))	//check if the variable is real( number with '.')
				endReal = true;
			else if (isNumber(buffer)) //check if the variable is number 
				endNumber = true;
			else {	//check if the variable is identifier
				if (!isdigit(buffer[0]))	// check if identifier not start with number
					endIdentifier = true;
			}

			currentIndex = 0;	// reset index for isNumber() and isReal()
		}
		print();
	}
}


int main() {
	string filename, outputFileName;
	cout << "Please enter file name where you want to read: ";
	cin >> filename;

	cout << "Please enter file name where you want to store output: ";
	cin >> outputFileName;

	
	fin.open(filename);
	outputFile.open(outputFileName);

	if (!fin.is_open()) {
		cout << "error while opening the file\n";
		exit(0);
	}
	cout << "TOKENS               Lexemes" << endl; // print table contents
	outputFile << "TOKENS               Lexemes\n";

	while (!fin.eof()) {
		ch = fin.get();
		lexer();	// include print() in the function
	}

	fin.close();	//close file
	outputFile.close(); //close output file

	system("pause");
	return 0;
}

File: pythonVersion/lexical_analyzer.py
operator = "*+-=/><%"
separator = "'(){}[],:;"
keywords = { "int", "float", "bool", "if", "else", "then", "endif", "while", "whileend", "do", "doend","for", "forend", "input", "output", "and", "or", "function" }

def isKeyword(buffer):
    for i in keywords:
        if i == buffer:
            return True
    return False

def isOperator(ch):
    for i in operator:
        if i == ch:
            return True
    return False

def isSeparator(ch):
    for i in separator:
        if i == ch:
            return True
    return False

def isNumber(buffer, currentIndex):
    count = 0
    while count < currentIndex:
        if not buffer[count].isdigit():
            return False
        count += 1
    return True

def isReal(buffer, currentIndex):
    str = convert(buffer)
    found = str.find(".")
    if found != -1:
        index = 0
        while index < found:
            if not str[index].isdigit():
                return false
            index += 1

        index = found + 1
        while index < currentIndex:
            if not str[index].isdigit():
                return false
            index += 1
        return True
    else:
        return False

def convert(buffer):
    new = ""
    for i in buffer:
        new += i
    return new




def main():
    isComments = False
    detection = [' '] * 2
    x = 0

    currentIndex = 0

    hasSep = False
    hasOpe = False

    endOperator = False
    endSeparator = False
    endNumber = False
    endReal = False
    endIdentifier = False

    buffer = ['\0'] * 15
    j = 0

    filename = input("Please enter file name where you want to read: ")
    outputFileName = input("Please enter file name where you want to store output: ")
    outputFile = open(outputFileName,'a')
    with open(filename) as fin:
        print("TOKENS               Lexemes")
        outputFile.write("TOKENS               Lexemes\n")
        while True:
            ch = fin.read(1)

            if not ch:
                fin.close()
                outputFile.close()
                exit(0)
            else:
                if ch == '!':
                    detection[x] = ch
                    x += 1
                    if detection[1] == ch:
                        isComments = False
                        x = 0
                        detection = '\0'
                    else:
                        isComments = True

                if not isComments:
                    if isSeparator(ch):
                        hasSep = True
                        endSeparator = True
                    if isOperator(ch):
                        hasOpe = True
                        endOperator = True
                    if ch.isalnum() or ch == '$' or ch == '.':
                        buffer[j] = ch
                        j += 1
                        currentIndex += 1
                    elif j != 0 and (ch == ' ' or ch == '\n' or hasSep or hasOpe or ch == None):
                        i = 0
                        while i < len(buffer):
                            if i >= j:
                                buffer[i] = '\0'
                            i += 1
                        j = 0
                        if isKeyword(buffer):
                            print("KEYWORD         =     " + convert(buffer))
                            outputFile.write("KEYWORD         =     " + convert(buffer) + "\n")
                        elif isNumber(buffer, currentIndex):
                            endNumber = True
                        elif isReal(buffer,currentIndex):
                            endReal = True
                        else:
                            if not buffer[0].isdigit():
                                endIdentifier = True
                        currentIndex = 0

                    if endIdentifier:
                        print("IDENTIFIER      =     " + convert(buffer))
                        outputFile.write("IDENTIFIER      =     " + convert(buffer)+ "\n")
                        endIdentifier = False
                    if endNumber:
                        print("NUMBER          =     " + convert(buffer))
                        outputFile.write("NUMBER          =     " + convert(buffer)+ "\n")
                        endNumber = False
                    if endReal:
                        print("REAL            =     " + convert(buffer))
                        outputFile.write("REAL            =     " + convert(buffer)+ "\n")
                        endReal = False
                    if endSeparator:
                        print("SEPARATOR       =     " + ch)
                        outputFile.write("SEPARATOR       =     " + ch + "\n")
                        endSeparator = False
                    if endOperator:
                        print("OPERATOR        =     " + ch)
                        outputFile.write("OPERATOR        =     " + ch + "\n")
                        endOperator = False



main()

